File: packages/roaster/src/roaster/findings_publication.py
"""Pure findings-publication helpers for roaster PR comments."""

from __future__ import annotations

import functools
import hashlib
import json
import re
from dataclasses import dataclass
from typing import Any, TypeAlias

from roaster.models import DiffLineLocation, ReviewFinding, TargetKind

_SEVERITY_LABELS: dict[str, str] = {
    "error": "⛔ error",
    "warning": "⚠️ warning",
    "info": "ℹ️ info",
}

_FOOTER = "_Post-only steelthread: this comment never blocks the check._"
_INLINE_MARKER_PREFIX = "roaster-inline"
_ACTIVITY_LOG_HEADING = "### Activity Log"
_ACTIVITY_LOG_CAP = 10


@dataclass(frozen=True)
class FindingsPayloadParseError:
    """Non-ideal parse result for malformed findings JSON input."""

    message: str
    error_type: str = "findings_parse_failed"


@dataclass(frozen=True)
class FindingsPublicationPolicyError:
    """Non-ideal result for valid findings payloads that cannot be published to a PR."""

    message: str
    error_type: str = "findings_publication_policy_failed"


@dataclass(frozen=True)
class InlinePostingStatusParseError:
    """Non-ideal parse result for malformed inline-posting JSON input."""

    message: str
    error_type: str = "inline_posting_parse_failed"


@dataclass(frozen=True)
class FindingsCommentBodyParseError:
    """Non-ideal parse result for a rendered findings comment body."""

    message: str
    error_type: str = "findings_comment_marker_parse_failed"


@dataclass(frozen=True)
class FindingsPayload:
    review_name: str
    base_ref: str
    count: int
    findings: tuple[ReviewFinding, ...]
    target_kind: TargetKind = "diff"
    target_label: str | None = None
    target_source_path: str | None = None
    error_type: str | None = None
    error_message: str | None = None

    @property
    def is_error(self) -> bool:
        return self.error_type is not None


@dataclass(frozen=True)
class InlinePostingStatus:
    posted_count: int
    skipped_duplicate_count: int
    fallback_only_count: int
    api_error: str | None = None


@dataclass(frozen=True)
class ParsedFindingsCommentBody:
    marker: str
    body: str


FindingsPayloadParseResult: TypeAlias = FindingsPayload | FindingsPayloadParseError
PublishableFindingsPayloadResult: TypeAlias = (
    FindingsPayload | FindingsPayloadParseError | FindingsPublicationPolicyError
)
InlinePostingStatusParseResult: TypeAlias = InlinePostingStatus | InlinePostingStatusParseError
FindingsCommentBodyParseResult: TypeAlias = (
    ParsedFindingsCommentBody | FindingsCommentBodyParseError
)


def parse_findings_payload_result(raw: str) -> FindingsPayloadParseResult:
    """Parse a roaster clinkr envelope, returning a payload or error object."""
    try:
        data = json.loads(raw)
    except json.JSONDecodeError as exc:
        return FindingsPayloadParseError(message=f"input is not valid JSON: {exc}")

    if not isinstance(data, dict):
        return FindingsPayloadParseError(
            message=f"expected a JSON object at top level, got {type(data).__name__}"
        )

    if "exit_code" not in data:
        return FindingsPayloadParseError(
            message="expected a clinkr envelope with top-level 'exit_code'"
        )

    exit_code = data.get("exit_code")
    if exit_code != 0:
        return FindingsPayload(
            review_name="unknown",
            base_ref="unknown",
            count=0,
            findings=(),
            error_type=_coerce_str(data.get("error_type"), default="unknown"),
            error_message=_coerce_str(data.get("message"), default=""),
        )

    inner = data.get("data")
    if not isinstance(inner, dict):
        return FindingsPayloadParseError(message="`data` must be an object when `exit_code` is 0")

    review_name = _coerce_str(inner.get("review_name"), default="unknown")
    base_ref = _coerce_str(inner.get("base_ref"), default="unknown")
    target_metadata = _parse_target_metadata(inner)
    if isinstance(target_metadata, FindingsPayloadParseError):
        return target_metadata
    target_kind, target_label, target_source_path = target_metadata

    raw_findings = inner.get("findings", [])
    if not isinstance(raw_findings, list):
        return FindingsPayloadParseError(message="`findings` must be a list when present")

    findings: list[ReviewFinding] = []
    for index, item in enumerate(raw_findings):
        parsed_finding = _parse_finding_result(item, index=index)
        if isinstance(parsed_finding, FindingsPayloadParseError):
            return parsed_finding
        findings.append(parsed_finding)

    count = inner.get("count")
    if not isinstance(count, int):
        count = len(findings)

    return FindingsPayload(
        review_name=review_name,
        base_ref=base_ref,
        count=count,
        findings=tuple(findings),
        target_kind=target_kind,
        target_label=target_label,
        target_source_path=target_source_path,
    )


def parse_inline_posting_status_result(raw: str) -> InlinePostingStatusParseResult:
    """Parse JSON emitted by ``roaster exec post-inline-findings``."""
    try:
        data = json.loads(raw)
    except json.JSONDecodeError as exc:
        return InlinePostingStatusParseError(message=f"inline result is not valid JSON: {exc}")

    if not isinstance(data, dict):
        return InlinePostingStatusParseError(
            message=f"expected inline result JSON object, got {type(data).__name__}"
        )

    status_data: Any = data
    if "data" in data:
        status_data = data["data"]
        if not isinstance(status_data, dict):
            return InlinePostingStatusParseError(message="inline result `data` must be an object")

    return _parse_inline_posting_status_object(status_data)


def parse_publishable_diff_findings_payload_result(raw: str) -> PublishableFindingsPayloadResult:
    """Parse findings JSON and reject valid local-only findings before PR publication."""
    payload = parse_findings_payload_result(raw)
    if isinstance(payload, FindingsPayloadParseError):
        return payload
    return ensure_publishable_diff_payload(payload)


def ensure_publishable_diff_payload(
    payload: FindingsPayload,
) -> FindingsPayload | FindingsPublicationPolicyError:
    """Reject local-only document review payloads before PR publication."""
    if payload.target_kind != "diff":
        return FindingsPublicationPolicyError(
            message=(
                "document review findings are local-output only "
                "and cannot be published as PR comments"
            )
        )
    for finding in payload.findings:
        if not isinstance(finding.location, DiffLineLocation):
            return FindingsPublicationPolicyError(
                message=(
                    "PR comment publication requires diff-line finding locations; "
                    f"got {finding.location.kind!r}"
                )
            )
    return payload


def render_findings_comment(
    payload: FindingsPayload,
    *,
    inline_status: InlinePostingStatus | None = None,
) -> str:
    """Render the payload as a Markdown comment body."""
    marker = summary_marker_for_review(payload.review_name)
    heading = f"## roaster · `{payload.review_name}`"
    lines: list[str] = [marker, heading, ""]

    if inline_status is not None:
        lines.extend(_render_inline_posting_status(inline_status))
        lines.append("")

    if payload.is_error:
        lines.extend(_render_error_body(payload))
    elif payload.count == 0:
        lines.append(f"**No findings** against base `{payload.base_ref}`. ✅")
    else:
        lines.extend(_render_findings_body(payload))

    return "\n".join(lines)


def summary_marker_for_review(review_name: str) -> str:
    return f"<!-- roaster:{review_name} -->"


def parse_findings_comment_body(raw: str) -> FindingsCommentBodyParseResult:
    lines = raw.splitlines()
    if not lines:
        return FindingsCommentBodyParseError(message="input body is empty")

    match = _get_summary_marker_pattern().match(lines[0].rstrip("\r"))
    if match is None:
        return FindingsCommentBodyParseError(
            message="first line of body must be a `<!-- roaster:<key> -->` marker"
        )

    return ParsedFindingsCommentBody(marker=f"<!-- {match.group(1)} -->", body=raw)


def preserve_activity_log(existing_body: str, new_body: str, run_summary: str) -> str:
    """Merge a new Activity Log entry into ``new_body`` from ``existing_body``."""
    prior_entries = _extract_activity_log_entries(existing_body)
    entries = list(prior_entries) + [run_summary]
    entries = entries[-_ACTIVITY_LOG_CAP:]

    base = _strip_activity_log(new_body).rstrip()
    rendered = [base, "", _ACTIVITY_LOG_HEADING, ""]
    rendered.extend(f"- {entry}" for entry in entries)
    return "\n".join(rendered) + "\n"


def inline_marker_for_finding(review_name: str, finding: ReviewFinding) -> str:
    location_discriminator = ""
    if not isinstance(finding.location, DiffLineLocation):
        location_discriminator = json.dumps(
            finding.location.model_dump(mode="json"),
            sort_keys=True,
            separators=(",", ":"),
        )
    digest_input = "\0".join(
        (
            review_name,
            finding.path or "",
            "" if finding.line is None else str(finding.line),
            location_discriminator,
            finding.severity,
            finding.summary,
            finding.details,
        )
    )
    digest = hashlib.sha256(digest_input.encode("utf-8")).hexdigest()[:16]
    return f"<!-- {_INLINE_MARKER_PREFIX}:{review_name}:{digest} -->"


def extract_inline_markers(body: str) -> tuple[str, ...]:
    markers: list[str] = []
    for line in body.splitlines():
        stripped = line.strip()
        if stripped.startswith(f"<!-- {_INLINE_MARKER_PREFIX}:") and stripped.endswith(" -->"):
            markers.append(stripped)
    return tuple(markers)


def render_inline_body(marker: str, finding: ReviewFinding, *, review_name: str) -> str:
    return "\n".join(
        [
            marker,
            f"**{finding.severity}: {finding.summary}**",
            f"_Review: `{review_name}`._",
            "",
            finding.details,
            "",
            "_Posted by roaster. Re-running may skip this comment by marker._",
        ]
    )


@functools.cache
def _get_summary_marker_pattern() -> re.Pattern[str]:
    return re.compile(r"^<!-- (roaster:[^ ]+) -->$")


def _render_inline_posting_status(status: InlinePostingStatus) -> list[str]:
    lines = [
        "### Inline posting",
        "",
        f"- **Inline comments posted:** {status.posted_count}",
        f"- **Duplicate inline comments skipped:** {status.skipped_duplicate_count}",
        f"- **Summary-only findings:** {status.fallback_only_count}",
    ]
    if status.api_error is not None:
        lines.append(f"- **API error:** {status.api_error}")
    return lines


def _render_error_body(payload: FindingsPayload) -> list[str]:
    return [
        f"**Roaster failed** against base `{payload.base_ref}`. ⚠️",
        "",
        f"- **Error type:** `{payload.error_type}`",
        f"- **Message:** {payload.error_message or '(none)'}",
    ]


def _render_findings_body(payload: FindingsPayload) -> list[str]:
    noun = "finding" if payload.count == 1 else "findings"
    body: list[str] = [
        f"**{payload.count} {noun}** against base `{payload.base_ref}`.",
        "",
        "| Severity | File | Line | Summary |",
        "| --- | --- | --- | --- |",
    ]
    body.extend(_finding_table_row(finding) for finding in payload.findings)
    body.extend(["", "<details>", "<summary>Details</summary>", ""])
    for finding in payload.findings:
        location = _finding_location(finding)
        body.extend(
            [
                f"### `{location}` — {finding.severity}",
                f"**{finding.summary}**",
                "",
                finding.details,
                "",
            ]
        )
    body.extend(["</details>", "", _FOOTER])
    return body


def _finding_table_row(finding: ReviewFinding) -> str:
    return (
        f"| {_severity_label(finding.severity)} | `{_path_display(finding.path)}` | "
        f"{_line_display(finding.line)} | {finding.summary} |"
    )


def _finding_location(finding: ReviewFinding) -> str:
    if finding.line is None:
        return _path_display(finding.path)
    return f"{_path_display(finding.path)}:{finding.line}"


def _severity_label(severity: str) -> str:
    return _SEVERITY_LABELS.get(severity, severity)


def _line_display(line: int | None) -> str:
    return "—" if line is None else str(line)


def _path_display(path: str | None) -> str:
    return "—" if path is None else path


def _coerce_str(value: Any, *, default: str) -> str:
    if isinstance(value, str) and value:
        return value
    return default


def _parse_target_metadata(
    inner: dict[str, Any],
) -> tuple[TargetKind, str | None, str | None] | FindingsPayloadParseError:
    target = inner.get("target")
    if target is not None:
        if not isinstance(target, dict):
            return FindingsPayloadParseError(message="`target` must be an object when present")
        raw_kind = target.get("kind", "diff")
        raw_label = target.get("label")
        raw_source_path = target.get("source_path")
    else:
        raw_kind = inner.get("target_kind", "diff")
        raw_label = inner.get("target_label")
        raw_source_path = inner.get("target_source_path")
    if raw_kind == "diff":
        target_kind: TargetKind = "diff"
    elif raw_kind == "document":
        target_kind = "document"
    else:
        return FindingsPayloadParseError(
            message="target kind must be either `diff` or `document` when present"
        )

    target_label = _optional_str(raw_label, field_name="target label")
    if isinstance(target_label, FindingsPayloadParseError):
        return target_label
    target_source_path = _optional_str(raw_source_path, field_name="target source_path")
    if isinstance(target_source_path, FindingsPayloadParseError):
        return target_source_path
    return target_kind, target_label, target_source_path


def _optional_str(value: Any, *, field_name: str) -> str | None | FindingsPayloadParseError:
    if value is None:
        return None
    if isinstance(value, str):
        return value or None
    return FindingsPayloadParseError(message=f"{field_name} must be a string when present")


def _parse_inline_posting_status_object(data: dict[str, Any]) -> InlinePostingStatusParseResult:
    posted_count = data.get("posted_count")
    skipped_duplicate_count = data.get("skipped_duplicate_count")
    fallback_only_count = data.get("fallback_only_count")
    api_error = data.get("api_error")

    if not isinstance(posted_count, int):
        return InlinePostingStatusParseError(message="inline result missing integer `posted_count`")
    if not isinstance(skipped_duplicate_count, int):
        return InlinePostingStatusParseError(
            message="inline result missing integer `skipped_duplicate_count`"
        )
    if not isinstance(fallback_only_count, int):
        return InlinePostingStatusParseError(
            message="inline result missing integer `fallback_only_count`"
        )
    if api_error is not None and not isinstance(api_error, str):
        return InlinePostingStatusParseError(
            message="inline result `api_error` must be a string or null"
        )

    return InlinePostingStatus(
        posted_count=posted_count,
        skipped_duplicate_count=skipped_duplicate_count,
        fallback_only_count=fallback_only_count,
        api_error=api_error,
    )


def _parse_finding_result(item: Any, *, index: int) -> ReviewFinding | FindingsPayloadParseError:
    if not isinstance(item, dict):
        return FindingsPayloadParseError(message=f"finding #{index} is not an object")

    try:
        return ReviewFinding.from_json_dict(item)
    except ValueError as exc:
        return FindingsPayloadParseError(message=f"finding #{index}: {exc}")


def _extract_activity_log_entries(body: str) -> tuple[str, ...]:
    lines = body.splitlines()
    for index, line in enumerate(lines):
        if line.strip() == _ACTIVITY_LOG_HEADING:
            entries: list[str] = []
            for entry_line in lines[index + 1 :]:
                stripped = entry_line.strip()
                if stripped.startswith("- "):
                    entries.append(stripped[2:])
            return tuple(entries)
    return ()


def _strip_activity_log(body: str) -> str:
    lines = body.splitlines()
    for index, line in enumerate(lines):
        if line.strip() == _ACTIVITY_LOG_HEADING:
            return "\n".join(lines[:index])
    return body

File: ts/packages/roaster/src/findings-publication.ts
import {
	buildFailureMachineEnvelopeSchema,
	buildSuccessMachineEnvelopeSchema,
	machineEnvelopeSchema,
} from "@asdl/clinkr";
import { formatZodError, truncatedSha256Digest } from "@asdl/core/primitives";
import { z } from "zod";

import type { RoasterContext } from "./context.ts";
import { classifyInlineFindings } from "./inline-commentability.ts";
import {
	reviewRunResultSchema,
	type InlinePostingStatus,
	type PRInlineCommentInput,
	type PostInlineFindingsResult,
	type ReviewFinding,
	type ReviewInputCoverage,
} from "./models.ts";

const SUMMARY_MARKER_RE = /^<!-- (roaster:[^ ]+) -->$/;
const INLINE_MARKER_PREFIX = "roaster-inline";
const ACTIVITY_LOG_HEADING = "### Activity Log";
const ACTIVITY_LOG_CAP = 10;
const OMITTED_INPUT_FILES_RENDER_LIMIT = 10;
const FOOTER = "_Posted by roaster. This comment is informational and does not block the check._";
const BOT_LOGIN = "github-actions[bot]";

const reviewRunSuccessEnvelopeSchema = buildSuccessMachineEnvelopeSchema(reviewRunResultSchema);

const reviewRunFailureEnvelopeSchema = buildFailureMachineEnvelopeSchema({
	errorTypeSchema: z.string().trim().min(1),
});

const SEVERITY_LABELS = {
	error: "⛔ error",
	warning: "⚠️ warning",
	info: "ℹ️ info",
} as const;

export interface FindingsPayload {
	readonly reviewName: string;
	readonly baseRef: string;
	readonly count: number;
	readonly findings: readonly ReviewFinding[];
	readonly inputCoverage: ReviewInputCoverage | null;
	readonly errorType: string | null;
	readonly errorMessage: string | null;
}

export interface FindingsPayloadParseError {
	readonly type: "findings_payload_parse_error";
	readonly message: string;
}

export interface FindingsCommentBodyParseError {
	readonly type: "findings_comment_body_parse_error";
	readonly message: string;
}

export interface ParsedFindingsCommentBody {
	readonly marker: string;
	readonly body: string;
}

export type FindingsPayloadParseResult =
	| { readonly type: "ok"; readonly payload: FindingsPayload }
	| { readonly type: "error"; readonly error: FindingsPayloadParseError };
export type FindingsCommentBodyParseResult =
	| { readonly type: "ok"; readonly parsed: ParsedFindingsCommentBody }
	| { readonly type: "error"; readonly error: FindingsCommentBodyParseError };

export function parseFindingsPayloadResult(
	raw: string,
	options: { readonly fallbackReviewName?: string; readonly fallbackBaseRef?: string } = {},
): FindingsPayloadParseResult {
	const fallbackReviewName = options.fallbackReviewName ?? "unknown";
	const fallbackBaseRef = options.fallbackBaseRef ?? "unknown";
	const data = parseJson(raw);
	if (data.type === "error") return payloadError(data.message);
	if (!machineEnvelopeSchema.safeParse(data.value).success)
		return payloadError("expected a clinkr envelope with top-level 'exit_code'");

	const success = reviewRunSuccessEnvelopeSchema.safeParse(data.value);
	if (success.success) {
		return {
			type: "ok",
			payload: {
				reviewName: success.data.data.reviewName,
				baseRef: success.data.data.baseRef,
				count: success.data.data.count,
				findings: success.data.data.findings,
				inputCoverage: success.data.data.inputCoverage,
				errorType: null,
				errorMessage: null,
			},
		};
	}

	const failure = reviewRunFailureEnvelopeSchema.safeParse(data.value);
	if (failure.success) {
		return {
			type: "ok",
			payload: {
				reviewName: fallbackReviewName,
				baseRef: fallbackBaseRef,
				count: 0,
				findings: [],
				inputCoverage: null,
				errorType: failure.data.error_type,
				errorMessage: failure.data.message,
			},
		};
	}

	return payloadError(`invalid review-run envelope: ${formatZodError(success.error)}`);
}

export interface PublishFindingsOptions {
	readonly prNumber: number;
	readonly envelope: string;
	readonly runUrl?: string | undefined;
	readonly fallbackReviewName?: string | undefined;
	readonly fallbackBaseRef?: string | undefined;
	readonly cwd: string;
	readonly env?: NodeJS.ProcessEnv | undefined;
	readonly signal?: AbortSignal | undefined;
}

export type PublishFindingsResult =
	| {
			readonly type: "ok";
			readonly inlineStatus: PostInlineFindingsResult;
			readonly summaryAction: "posted" | "updated";
	  }
	| { readonly type: "error"; readonly message: string };

export async function publishFindings(
	ctx: Pick<RoasterContext, "github">,
	options: PublishFindingsOptions,
): Promise<PublishFindingsResult> {
	const parsed = parseFindingsPayloadResult(options.envelope, fallbackPayloadOptions(options));
	if (parsed.type === "error") return publicationError(parsed.error.message);

	const inlineStatus = await postInlineFindings(ctx, parsed.payload, options);
	const renderedBody = renderFindingsComment(parsed.payload, { inlineStatus });
	const parsedBody = parseFindingsCommentBody(renderedBody);
	if (parsedBody.type === "error") return publicationError(parsedBody.error.message);

	const existing = await ctx.github.findPrDiscussionCommentByMarker({
		prNumber: options.prNumber,
		marker: parsedBody.parsed.marker,
		authorLogin: BOT_LOGIN,
		...githubOptions(options),
	});
	if (existing.type === "error") return publicationError(existing.error.message);

	const nextBody = preserveActivityLog(
		existing.value?.body ?? "",
		parsedBody.parsed.body,
		activityLogEntry(options.runUrl),
	);
	const written =
		existing.value === null
			? await ctx.github.addPrDiscussionComment(options.prNumber, nextBody, githubOptions(options))
			: await ctx.github.updatePrDiscussionComment(
					existing.value.id,
					nextBody,
					githubOptions(options),
				);
	if (written.type === "error") return publicationError(written.error.message);

	return {
		type: "ok",
		inlineStatus,
		summaryAction: existing.value === null ? "posted" : "updated",
	};
}

export function renderFindingsComment(
	payload: FindingsPayload,
	options: { readonly inlineStatus?: InlinePostingStatus | null | undefined } = {},
): string {
	const lines = [
		summaryMarkerForReview(payload.reviewName),
		`## roaster · \`${payload.reviewName}\``,
		"",
	];
	if (options.inlineStatus !== undefined && options.inlineStatus !== null) {
		lines.push(...renderInlinePostingStatus(options.inlineStatus), "");
	}
	if (payload.errorType !== null) {
		lines.push(...renderErrorBody(payload));
	} else {
		if (payload.inputCoverage !== null)
			lines.push(...renderInputCoverage(payload.inputCoverage), "");
		lines.push(
			...(payload.count === 0 ? renderNoFindingsBody(payload) : renderFindingsBody(payload)),
		);
	}
	return lines.join("\n");
}

export function summaryMarkerForReview(reviewName: string): string {
	return `<!-- roaster:${reviewName} -->`;
}

export function parseFindingsCommentBody(raw: string): FindingsCommentBodyParseResult {
	const firstLine = raw.split(/\r?\n/u)[0];
	if (firstLine === undefined || firstLine === "") return commentBodyError("input body is empty");
	const match = SUMMARY_MARKER_RE.exec(firstLine.trimEnd());
	if (match === null)
		return commentBodyError("first line of body must be a `<!-- roaster:<key> -->` marker");
	return { type: "ok", parsed: { marker: `<!-- ${match[1]} -->`, body: raw } };
}

export function inlineMarkerForFinding(reviewName: string, finding: ReviewFinding): string {
	const digestInput = [
		reviewName,
		finding.path ?? "",
		finding.line === null ? "" : String(finding.line),
		finding.severity,
		finding.summary,
		finding.details,
	].join("\0");
	const digest = truncatedSha256Digest(digestInput).slice(0, 16);
	return `<!-- ${INLINE_MARKER_PREFIX}:${reviewName}:${digest} -->`;
}

export function extractInlineMarkers(body: string): readonly string[] {
	return body
		.split(/\r?\n/u)
		.map((line) => line.trim())
		.filter((line) => line.startsWith(`<!-- ${INLINE_MARKER_PREFIX}:`) && line.endsWith(" -->"));
}

export function renderInlineBody(
	marker: string,
	finding: ReviewFinding,
	options: { readonly reviewName: string },
): string {
	return [
		marker,
		`**${finding.severity}: ${finding.summary}**`,
		`_Review: \`${options.reviewName}\`._`,
		"",
		finding.details,
		"",
		"_Posted by roaster. Re-running may skip this comment by marker._",
	].join("\n");
}

export function preserveActivityLog(
	existingBody: string,
	newBody: string,
	runSummary: string,
): string {
	const entries = [...extractActivityLogEntries(existingBody), runSummary].slice(-ACTIVITY_LOG_CAP);
	return (
		[
			stripActivityLog(newBody).trimEnd(),
			"",
			ACTIVITY_LOG_HEADING,
			"",
			...entries.map((entry) => `- ${entry}`),
		].join("\n") + "\n"
	);
}

async function postInlineFindings(
	ctx: Pick<RoasterContext, "github">,
	payload: FindingsPayload,
	options: PublishFindingsOptions,
): Promise<PostInlineFindingsResult> {
	if (payload.errorType !== null || payload.count === 0) return emptyInlineResult();

	let changedFilesResult: Awaited<ReturnType<RoasterContext["github"]["getPrChangedFiles"]>>;
	try {
		changedFilesResult = await ctx.github.getPrChangedFiles(
			options.prNumber,
			githubOptions(options),
		);
	} catch (caught) {
		return { ...emptyInlineResult(), apiError: caughtMessage(caught) };
	}
	if (changedFilesResult.type === "error") {
		return { ...emptyInlineResult(), apiError: changedFilesResult.error.message };
	}

	let reviewCommentsResult: Awaited<ReturnType<RoasterContext["github"]["getPrReviewComments"]>>;
	try {
		reviewCommentsResult = await ctx.github.getPrReviewComments(
			options.prNumber,
			githubOptions(options),
		);
	} catch (caught) {
		return { ...emptyInlineResult(), apiError: caughtMessage(caught) };
	}
	if (reviewCommentsResult.type === "error") {
		return { ...emptyInlineResult(), apiError: reviewCommentsResult.error.message };
	}

	const classified = classifyInlineFindings(payload.findings, changedFilesResult.value);
	const existingMarkers = new Set(
		reviewCommentsResult.value
			.filter((comment) => comment.author === BOT_LOGIN)
			.flatMap((comment) => extractInlineMarkers(comment.body)),
	);
	const comments: PRInlineCommentInput[] = [];
	let skippedDuplicateCount = 0;

	for (const item of classified.inlineable) {
		const marker = inlineMarkerForFinding(payload.reviewName, item.finding);
		if (existingMarkers.has(marker)) {
			skippedDuplicateCount += 1;
			continue;
		}
		comments.push({
			path: item.target.path,
			line: item.target.line,
			body: renderInlineBody(marker, item.finding, { reviewName: payload.reviewName }),
		});
	}

	let apiError: string | null = null;
	let postedCount = 0;
	if (comments.length > 0) {
		try {
			const posted = await ctx.github.createPrReview(
				options.prNumber,
				comments,
				githubOptions(options),
			);
			if (posted.type === "error") apiError = posted.error.message;
			else postedCount = comments.length;
		} catch (caught) {
			apiError = caughtMessage(caught);
		}
	}

	return {
		postedCount,
		skippedDuplicateCount,
		fallbackOnlyCount: classified.fallbackOnly.length,
		apiError,
		fallbackOnly: classified.fallbackOnly,
	};
}

function emptyInlineResult(): PostInlineFindingsResult {
	return {
		postedCount: 0,
		skippedDuplicateCount: 0,
		fallbackOnlyCount: 0,
		apiError: null,
		fallbackOnly: [],
	};
}

function fallbackPayloadOptions(
	options: Pick<PublishFindingsOptions, "fallbackReviewName" | "fallbackBaseRef">,
): {
	readonly fallbackReviewName?: string;
	readonly fallbackBaseRef?: string;
} {
	return {
		...(options.fallbackReviewName === undefined
			? {}
			: { fallbackReviewName: options.fallbackReviewName }),
		...(options.fallbackBaseRef === undefined ? {} : { fallbackBaseRef: options.fallbackBaseRef }),
	};
}

function githubOptions(options: Pick<PublishFindingsOptions, "cwd" | "env" | "signal">): {
	readonly cwd: string;
	readonly env?: NodeJS.ProcessEnv | undefined;
	readonly signal?: AbortSignal | undefined;
} {
	return {
		cwd: options.cwd,
		...(options.env === undefined ? {} : { env: options.env }),
		...(options.signal === undefined ? {} : { signal: options.signal }),
	};
}

function activityLogEntry(runUrl: string | undefined): string {
	const timestamp = new Date().toISOString();
	return runUrl === undefined || runUrl.trim() === "" ? timestamp : `${timestamp} · ${runUrl}`;
}

function caughtMessage(caught: unknown): string {
	return caught instanceof Error ? caught.message : String(caught);
}

function publicationError(message: string): PublishFindingsResult {
	return { type: "error", message };
}

function renderInlinePostingStatus(status: InlinePostingStatus): string[] {
	const lines = [
		"### Inline posting",
		"",
		`- **Inline comments posted:** ${status.postedCount}`,
		`- **Duplicate inline comments skipped:** ${status.skippedDuplicateCount}`,
		`- **Summary-only findings:** ${status.fallbackOnlyCount}`,
	];
	if (status.apiError !== null) lines.push(`- **API error:** ${status.apiError}`);
	return lines;
}

function renderErrorBody(payload: FindingsPayload): string[] {
	return [
		`**Roaster failed** against base \`${payload.baseRef}\`. ⚠️`,
		"",
		`- **Error type:** \`${payload.errorType ?? "unknown"}\``,
		`- **Message:** ${payload.errorMessage ?? "(none)"}`,
	];
}

function renderInputCoverage(coverage: ReviewInputCoverage): string[] {
	const lines = [
		"### Review input coverage",
		"",
		`- **Filtered diff files:** ${coverage.changedPathCount} (${coverage.includedFileCount} supplied to bounded prompt input, ${coverage.omittedFileCount} omitted)`,
		`- **Filtered diff estimate:** ~${coverage.fullDiffEstimatedTokens} tokens`,
		`- **Prompt caps:** ${coverage.promptDiffTokenCap} diff tokens total; ${coverage.promptDiffFileTokenCap} tokens per file diff`,
	];
	if (coverage.omittedFileCount === 0)
		return [
			...lines,
			"",
			"Bounded prompt input included all files in the filtered diff after configured roaster exclusions.",
		];
	lines.push(
		"",
		"Review completed with bounded prompt input. The following filtered-diff file segments were not supplied in prompt input after configured roaster exclusions:",
	);
	for (const file of coverage.omittedFiles.slice(0, OMITTED_INPUT_FILES_RENDER_LIMIT)) {
		lines.push(
			`- \`${file.path}\` (${file.changeKind}, ${file.byteSize} bytes, ~${file.estimatedTokens} tokens, +${file.addedLines}/-${file.removedLines}; ${file.reason.replaceAll("_", " ")})`,
		);
	}
	const remaining =
		coverage.omittedFileCount -
		Math.min(coverage.omittedFiles.length, OMITTED_INPUT_FILES_RENDER_LIMIT);
	if (remaining > 0) lines.push(`- …and ${remaining} more omitted file diff(s).`);
	return lines;
}

function renderNoFindingsBody(payload: FindingsPayload): string[] {
	if (payload.inputCoverage !== null && payload.inputCoverage.omittedFileCount > 0)
		return [
			`**No findings in the reviewed bounded input** against base \`${payload.baseRef}\`. ✅`,
		];
	return [`**No findings** against base \`${payload.baseRef}\`. ✅`];
}

function renderFindingsBody(payload: FindingsPayload): string[] {
	const noun = payload.count === 1 ? "finding" : "findings";
	const lines = [
		`**${payload.count} ${noun}** against base \`${payload.baseRef}\`.`,
		"",
		"| Severity | File | Line | Summary |",
		"| --- | --- | --- | --- |",
	];
	lines.push(...payload.findings.map(findingTableRow));
	lines.push("", "<details>", "<summary>Details</summary>", "");
	for (const finding of payload.findings) {
		lines.push(
			`### \`${findingLocation(finding)}\` — ${finding.severity}`,
			`**${finding.summary}**`,
			"",
			finding.details,
			"",
		);
	}
	lines.push("</details>", "", FOOTER);
	return lines;
}

function findingTableRow(finding: ReviewFinding): string {
	return `| ${SEVERITY_LABELS[finding.severity]} | \`${finding.path ?? "—"}\` | ${lineDisplay(finding.line)} | ${finding.summary} |`;
}

function findingLocation(finding: ReviewFinding): string {
	const path = finding.path ?? "unknown path";
	return finding.line === null ? path : `${path}:${finding.line}`;
}

function lineDisplay(line: number | null): string {
	return line === null ? "—" : String(line);
}

function extractActivityLogEntries(body: string): readonly string[] {
	const lines = body.split(/\r?\n/u);
	const headingIndex = lines.findIndex((line) => line.trim() === ACTIVITY_LOG_HEADING);
	if (headingIndex === -1) return [];
	return lines
		.slice(headingIndex + 1)
		.map((line) => line.trim())
		.filter((line) => line.startsWith("- "))
		.map((line) => line.slice(2));
}

function stripActivityLog(body: string): string {
	const lines = body.split(/\r?\n/u);
	const headingIndex = lines.findIndex((line) => line.trim() === ACTIVITY_LOG_HEADING);
	return headingIndex === -1 ? body : lines.slice(0, headingIndex).join("\n");
}

type JsonResult =
	| { readonly type: "ok"; readonly value: unknown }
	| { readonly type: "error"; readonly message: string };

function parseJson(raw: string): JsonResult {
	try {
		return { type: "ok", value: JSON.parse(raw) };
	} catch (caught) {
		return {
			type: "error",
			message: `input is not valid JSON: ${caught instanceof Error ? caught.message : String(caught)}`,
		};
	}
}

function payloadError(message: string): FindingsPayloadParseResult {
	return { type: "error", error: { type: "findings_payload_parse_error", message } };
}

function commentBodyError(message: string): FindingsCommentBodyParseResult {
	return { type: "error", error: { type: "findings_comment_body_parse_error", message } };
}
